### ISSUE TITLE
Proposal: Virtual extension methods

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#258]
Add support for _virtual extension methods_ - methods in interfaces with concrete implementations. A class that implements such an interface is required to have a single _most specific_ implementation for the interface method inherited from its base classes or interfaces. Virtual extension methods enable an API author to add methods to an interface in future versions without breaking source or binary compatibility with existing implementations of that interface.

These are similar to Java's ["Default Methods"](http://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html).

This feature requires corresponding support in the CLI/CLR. Programs that take advantage of this feature cannot run on earlier versions of the platform.

**Comment by thomaslevesque** [comment]
I really liked the idea of default methods when this feature was announced for Java 8. In some ways, it solves the same problem as extension methods, but it have one advantage and one drawback:

- **Pro**: the method can be overriden by derived classes to provide a more specific implementation
- **Con**: you can't extend interfaces you don't own

If C# could have default methods in addition to extension methods, it would be quite awesome. The first use case that comes to mind is a `Sort` method for `IList<T>`.

**Comment by theraot** [comment]
Extension method are pretty much that, in the sense that the compiler will use an actual method before looking for extension methods. So you can have that `Sort` method for `IList<T>` already.

---

On the other hand, these "virtual extension method" as described by @gafter are meant to serve as default implementations of the interface methods.

---

I'll call them vem for the purpose of this post, ok? ok...

Consider if you want to be able to declare an interface with a vem for some or all of it's members, and you want to be able to share these with other assemblies. Ok, that means that those implementations got to be stored in the assembly somewhere...

Since part the intention is that you will be able to use an assembly compiled against a version of an interface that had less methods than the current one, and also have the current one fill-in the missing methods... a more clever virtual resolution will be required, since the objects exported by the assembly don't have the methods you want - nor are aware of their existence.

In that situation we know that the missing methods will only be called from new code (since old code is not aware of them anyway). So it is sensible to store these vems is to implement them in similar ways to extension methods.  maybe using another attribute that the compiler knows about, so the compiler should be able to find them to generate the calls - these calls are resolved at compile time.

---

Now... you would also want to define these vems for interfaces that you did not declare (for example `IList<T>`), so we will need a syntax to specify these without creating a new interface... (allowing multiple vems per interface) and then how does the compiler decide which one to use?

- The first solution is to give identifiers to these vems. Have the programmer declare and use them by name. If none is given, it could fallback to...
- The next solution is is scoping (have it use the more local available, by namespace at least). If none, it could fallback to...
- The final solution is to allow only one per assembly and have it bound via assembly attributes, if the current assembly doesn't have one, look for one in the assembly that declared the interface.

---

If it is allowed to declare these vems without declaring the interface at the same time (I mean, if it is allowed to declare these vems for an interface declared somewhere else), that means that:
- Similar syntax to declaring a class or an interface is needed for vems.
- If the vems are named, two things happen:
  - Syntax to use them when declaring a class, will be needed too.
  - It is conceivable to have vems without interfaces. That would make them proper mixins.

---

Consider the case where you have multiple interface inheritance and various vems causing collisions, it could be solved by:
- Making such collision an error, and have you write the code when that happens.
- A mechanism similar to explicit implementation, it would be a syntax to say: I'm implementing this member using this vem (by name).

---

All in all, allowing properties and fields is problematic: if I implement the property replacing the one imported via vem, any other method in the vem that uses the backing field of that property (and that was not replaced) will not work as expected... so either only methods, or no partial replacements.

**Comment by thomaslevesque** [comment]
> Extension method are pretty much that, in the sense that the compiler will use an actual method before looking for extension methods. So you can have that Sort method for IList<T> already.

No, you can't; extension methods are resolved at compile time, not at runtime. So if you have a variable of type `IList<T>`, and you call a `Sort` extension method on it, it won't matter if the concrete type actually has a `Sort` instance method; it will never be called, since the compiler will already have transformed the call to a static method call.

Now, if the variable is known at compile time to be `List<T>`, it's a different matter; the compiler will indeed pick the instance method.

**Comment by theraot** [comment]
Ok... so the code has to be able to decide if the `IList<T>` is actually a `List<T>` (or other implementation), and do so at runtime because there is no way to know all the implementations beforehand, and still have intellisense...

So... the compiler would generate a method that will query the type, decide to use either the specific method if available or the default implementation, then the calls are resolved to this method. Then at runtime it is decided whatever or not to use the default implementation, when writing code you would have intellisense because the compiler knows the method will be available.

Note: That  hopefully could be optimized for known implementations. 

I can see how that can be implemented via libraries (It needs memorization, so the check is done only once per type per AppDomain), There will be a small penalty on each call due to the indirection, and a slightly bigger on the first call due to the check. Then it would be a matter of adding syntactic sugar to it.

What implementation to fallback to, can be decided at compile time, but whatever it is used or not, is decided at runtime. That implementation can be chosen beforehand because we know that if any new method is added in the dependency assembly it is not used in old code. The code of the default implementation must be referenced (cannot be copied to the current assembly) because they may change from a version to another of the dependency assembly.

**Comment by gafter** [comment]
@theraot What you're describing does not make sense to me. In this proposal the VM would use a vtable to cache the method to be invoked, just as it does today for interfaces. The CLR's method resolution mechanism for these methods would be enhanced to look through interface implementations if there is no implementation in the class hierarchy of the instance type. This is exactly what Java does today.

**Comment by theraot** [comment]
@gafter hmmm... seems to be you are trying to solve something different. What I gather is that you start from that these are members of the interface. I did start from that these are extension methods, having them to implement interface members being a secondary thing.

If you start by thinking about extension methods, and go on to make them "virtual" as to allow what @thomaslevesque proposes of extending an already existing interface (`IList<T>`) to provide a genal and default implementation of a method (`Sort`) [up to there extension methods will do], but have it use the specific implementation based on type of the object at runtime [that extension methods don't do], then you will end up with a different solution.

And, in reality if the API author wants to add new methods to existing interfaces and not break binary nor source, they can add extension methods. The only difference is that we want to use the specific implementation if available.

I suspect it would not be much different implementation, although I was imagining it more like dynamic...

Is the compiler able to pre-generate these vtable caches for known implementations at compile time? If no... then that's a suggestion.

**Comment by HaloFour** [comment]
@gafter I've love to hear comments/concerns regarding how C# might mitigate multiple inheritance concerns that arise from this.

For example:

```cs
public interface IFoo {
    void M() { }
}

public interface IBar {
    void M() { }
}

public class Baz : IFoo, IBar { }

Baz baz = new Baz();
baz.M();
```

In Java this is illegal and `Baz` is required to implement `M()`.  But of course `M()` could be added to the interface after the fact in which case `Baz` wouldn't supply an implementation.

**Comment by gafter** [comment]
@HaloFour I suspect the Java runtime method resolution algorithm would make sense for CLR/C# as well.

**Comment by fschmied** [comment]
In order to preserve binary compatibility when new members with a default implementation are added to existing interfaces, virtual extension methods do need CLR support.

However, if CLR support were not available, virtual extension methods could also be implemented as a language-only syntactic sugar feature, see #73. This could be seen as a "version 1" variant of the feature (with CLR support coming later, for example) that only provides source code compatibility, but not binary compatibility.

Also, virtual extension methods are not only about adding new members to existing interfaces; they are also about providing default implementations for interface members commonly (but not always) implemented in a certain way. (E.g., `IEnumerable<T>` providing a default implementation for the non-generic `IEnumerable` members which just delegate to the generic ones.)

**Comment by HaloFour** [comment]
The problem with that implementation is that it doesn't provide the biggest benefit to actual default implementations, that is being able to amend an existing interface with new members without breaking any existing code.  If it was handled as syntax candy entirely by the compiler then any code implementing that interface would require to be recompiled.

I'd rather see this implemented "properly" with CLR support, assuming that the CLR could be amended to support it within the C# 7.0 time frame.  If that's not the case then maybe consider the alternatives, but even then *my* preference might just be to hold off on the language feature until the CLR supports it.

**Comment by fschmied** [comment]
@HaloFour I agree that CLR support would be best, of course. However, I disagree that without the binary compatibility (which can only be provided by the CLR), the biggest benefit is lost.

First, to me as a framework auther and a framework user, the biggest benefit is that of having **overridable default implementations** for interface members that are **most often, but not always, implemented the same way**. This would
* allow me to "mix in" overridable behavior into my class simply by implementing an interface containing virtual extension methods,
* allow me to provide overridable default implementations for convenience methods (see e.g., LINQ operators - wouldn't it be nice if an implementation of `IEnumerable<T>` could override `Count` and `Any` to have better performance, for example?), and
* allow me to provide default implementations for methods inherited from base interfaces (see e.g.,  `IEnumerable<T>`, where the base `IEnumerable.GetEnumerator()` method is always implemented by delegating to the generic one).

Second, as a framework or library author, the feature of being able to add members to existing interfaces without breaking source code compatibility would _still_ be useful even if recompilation of implementers was required. While binary backward compatibility is really important to people implementing the .NET BCL, it is not as important to people implementing other libraries. A breaking change just requiring recompilation is still much better than a breaking change requiring modification of all implementers.

Therefore, in my opinion, the feature would be *better* with, but still **very valuable even without CLR support**. And thus, I would not want it to be postponed just because the CLR can't be changed by the time C# 7 is released.

**Comment by vladd** [comment]
Dear all, isn't this functionality just a simpler version of mixins? If we are considering having mixins in the language, we don't need to just copy a feature from other language.
The motivation from Java tutorial doesn't seem to be very convincing for me: changing the interface which is implemented by other's classes is anyway a breaking change, and having the other's code silently compiling without forcing the developers to review it is potentially dangerous.
Besides that, confirming to an interface and sharing the implementation should be potentially orthogonal things, right? With mixins we would add a possibility to opt in for code sharing while requiring the interface implementation. With the proposed feature one needs to opt out of the code sharing.

**Comment by gafter** [comment]
@vladd The problematic issue with multiple inheritance is state. But unlike mixins, this proposal does not allow the addition of state to interfaces. The default method implementation provided in an interface would be a correct (if inefficient) one based on other methods of the interface. This is like the existing extension method feature but these methods are virtual and can be overridden in subclasses for efficiency.

**Comment by vladd** [comment]
@gafter 
My main point is that extending the interface is **indeed** a breaking change, so making it undercover (so that the developers won't have to adjust their code) is not what I would endorse.

Having said that, I am not proposing adding state to _interfaces_. I think the needed functionality could be provided with mixins by the code like this (speculative syntax):

(old interface)
```csharp
interface I { TimeSpan ExpectedDuration { get; } }
```
(new interface)
```csharp
interface I
{
    TimeSpan ExpectedDuration { get; }
    TimeSpan ActualDuration { get; }
}

mixin ActualDurationImpl(T) where T : I
{
    public TimeSpan ActualDuration { get { return this.ExpectedDuration; } }
}
```
so that the implementers of the `I` interface would have an easy way to conform to the updated interface, but _not have this implicitly_!

**Comment by gafter** [comment]
@vladd In what sense is adding a virtual extension method a breaking change? The whole point of this feature is that adding them is *not* a breaking change.

**Comment by vladd** [comment]
@gafter Well, what if a class has already a method named like the one being added (but has different semantics)? What if the added method just doesn't fit into the semantics of the class implementing the original interface, so now it would be impossible to implement the updated interface? Say, if we had `interface IDrink { void Consume(); }`, `class Poison : IDrink { ... }`, and now adding `bool DoesTasteGood()` to `IDrink`?
This all **must** be a breaking semantical change, if even it's technically possible to make the code compilable.

**Comment by HaloFour** [comment]
@vladd 

> Well, what if a class has already a method named like the one being added (but has different semantics)?

If the class already had the method and that class was in a different assembly that was not recompiled after the interface was changed then nothing would change.  The method on the class would effectively shadow the interface method and callers to that method would call the class method directly.  Any new code written to call the interface method would call that method directly since the class method does not implement it.

If the assembly containing the class is then recompiled the compiler would consider the class method to implement the interface method and going forward calls to either would call the class method.

> What if the added method just doesn't fit into the semantics of the class implementing the original interface, so now it would be impossible to implement the updated interface?

C# has implicit implementation, this is not really a problem like it is in Java.

As for the semantics argument, you could pretty much invent scenarios to invalidate every feature proposed by claiming that the semantics can't be guaranteed.  I recall a pro-Java article written shortly after C# was first released claiming that delegates were stupid because of this very reason.  Apparently a `FireDelegate` delegate was less descriptive than an `IFireable` interface and you could accidentally launch a missile or something silly.

* To note, I hate this "virtual extension method" name.  The "extension" moniker simply doesn't fit.  I know that we probably don't want to use Java's naming convention of "default methods" because, ya know, ew Java, but we shouldn't adopt their habit of giving things wacky names to avoid the obvious name that someone else might've already used.

**Comment by vladd** [comment]
@HaloFour 
> If the assembly containing the class is then recompiled the compiler would consider the class method to implement the interface method and going forward calls to either would call the class method.
Exactly, and that's my point. This is dangerous. Your class now overrides the default method implementation with some arbitrary code, without the conscious decision of the developer, and without any warnings on recompilation.

**Comment by vladd** [comment]
@HaloFour 
> As for the semantics argument, you could pretty much invent scenarios to invalidate every feature proposed by claiming that the semantics can't be guaranteed.

No, my point was that adding a new interface method is a breaking semantical change, and as such ought to be a breaking syntactical change as well.

**Comment by thomaslevesque** [comment]
> Well, what if a class has already a method named like the one being added (but has different semantics)?

This problem is easily addressed : make it mandatory to should explicitly `override` the interface method (same as when you override a method from an abstract class). If you don't, then the method already in the class hides the one from the interface (as if you had added the `new` modifier), and it won't be called through the interface.

**Comment by vladd** [comment]
@thomaslevesque ... unless the existing method already `override`s some other interface's method?

**Comment by gafter** [comment]
@VSadov just pointed out to me that although it would be nice to add to `IEnumerator<T>`

```cs
    bool TryGetNext(out T result)
    {
        if (MoveNext())
        {
            result = Current;
            return true;
        }
        else
        {
            result = default(T);
            return false;
        }
    }
```

We cannot because the rules of covariance do not allow even an `out` parameter to be of type `T`. That's because it is represented in the CLR as a `ref` parameter.

**Comment by HaloFour** [comment]
@gafter Yeah, you'd need to return another covariant interface:

```cs
public interface IEnumerated<out T>
{
    bool Success { get; }
    T Current { get; }
}
```

Shame variance cannot be applied to structs, would be a lot cleaner.

**Comment by jveselka** [comment]
What if these virtual extension methods were explicitly implemented interface methods. They could still be overriden in implementations (but again only as explicit interface imlementation) and it would avoid problems @vladd pointed out.

**Comment by gafter** [comment]
Possibly a dup of #73.

**Comment by gafter** [comment]
See #73 for a detailed description.

**Comment by orthoxerox** [comment]
I wonder if the name for this concept was chosen well. To me a virtual extension method is literally an extension method that is virtual, i.e., something to solve the expression problem with, e.g.:

```c#
public class Base
{
    public virtual int FooVirt() => 1;
    
    public int FooStat() => -1;
}

public class Derived : Base
{
    public override int FooVirt() => 2;
    
    public int FooStat() => -2;
}

public static class Ext
{
    public static virtual int BarVirt(this Base @this) => 3;
    
    public static virtual int BarVirt(this Derived @this) => 4;
    
    public static int BarStat(this Base @this) => -3;
    
    public static int BarStat(this Derived @this) => -4;
    
    public static void Test()
    {
        Base b = new Derived();
        Console.WriteLine(b.FooVirt()); //2
        Console.WriteLine(b.FooStat()); //-1
        Console.WriteLine(b.BarVirt()); //4
        Console.WriteLine(b.BarStat()); //-3
    }
}
```

**Comment by alrz** [comment]
@gafter If we ever have unified nullables couldn't it be like
```
default T? Next() {
  return MoveNext() ? Current : null;
}
```
As long as `T` is not nullable?

Also with this we can define a non-generic method on top of generic interfaces, like
```cs
interface IEnumerable<out T> : IEnumerable {
  default IEnumerator IEnumerable.GetEnumerator() => this.GetEnumerator();
  IEnumerator<T> GetEnumerator();
}
interface IEnumerator<out T> : IDisposable, IEnumerator {
  default object IEnumerator.Current => (object) this.Current;
  T Current { get; }
}
```
which is great!

**Comment by MouseProducedGames** [comment]
This seems less a default, and more of a fallback. Ie., something to use if the class doesn't implement it.

Perhaps a name something like "fallback interface method" might work better. Except for "fall" looking like "fail".

...Maybe a different name.

**Comment by timgoodman** [comment]
Is there any chance we could get a syntactic sugar for implementation-through-composition, like Jon Skeet suggests in his answer here?
http://stackoverflow.com/questions/255553/is-it-possible-to-implement-mixins-in-c

That would allow you to support default methods (putting the default implementation in its own class, e.g. DefaultIEnumerator) which the consumer of the interface opts in to by adding a single line of code to their class.  It also lets the consumer of the interface choose a different implementation as their default.  Plus it's useful for doing composition on your own classes.  And it's really just a syntactic sugar, no need for CLR changes.

```
public class Foo : ISomeInterface
{
    private DefaultISomeInterface impl implements ISomeInterface; // new keyword: implements

    public void OneMethodOfISomeInterface()
    {
        // Specialize just this method
    }
}
```

**Comment by gafter** [comment]
This proposal is now tracked at https://github.com/dotnet/csharplang/issues/52
